### DISCUSSION TITLE
What does a Scoped Service Lifetime mean to MAUI

### DISCUSSION THREAD

**Discussion by michaelstonis** [dotnet/maui#7671] [Q&A]
The `MauiAppBuilder.Services` property has a number of ways to register objects with dependency injection. 

`AddTransient` will create a new instance of the object each time it is requested. 

`AddSingleton` will return a single/reusable instance of an object. 

`AddScoped`, in the context of ASP.NET will create a new instance per HTTP request. This paradigm doesn't exactly sync up with MAUI, with the exception of Blazor apps hosted in a MAUI container.

Is there a defined use or meaning of a `Scoped` service in non-blazor MAUI?

**Comment by Houri93** [comment]
this does require some clarification in terms of MAUI and MAUI Blazor, what is the scope of a scoped service, where does it start and end?

**Comment by aritchie** [comment]
Scoped services last for the lifetime of a page where they are resolved

**Reply by Houri93** [reply]
> Scoped services last for the lifetime of a page where they are resolved

Blazor apps are single page applications, does that mean they are scoped for the lifetime of the page which is the whole app on the client?

in Blazor WASM, what would be the difference between singleton and scoped then since each tab is an app instance? would the same logic apply to MAUI Blazor?

**Comment by aritchie** [comment]
The original question was for non-blazor MAUI.  

Any scoped services will hang around with the scope of a MAUI page.  If you stay completely within a blazor MAUI, then yes - those instances will essentially be singleton.

I don't know the answer for Blazor WASM.  Most services I make are singleton.

**Reply by goatrodeosoftware** [reply]
I don't think this is incorrect- if you put breakpoints in constructors of scoped services, they're called as you navigate, not just once on startup.

**Comment by hugebug4ever** [comment]
From the [official document](https://learn.microsoft.com/en-us/dotnet/architecture/maui/dependency-injection), it's not providing any information about scoped serice.

For some design purpose, I tried to use scoped service for AppShell, but it seems the scope is not populated to it's child (ContentTemplate). I take it as an design defect for "MAUI with Dependency Injection".

**Comment by Vroomer** [comment]
Scoped services can be useful in MAUI. For example if you have a page with lot's of ContentViews that use their own ViewModels and you need to share injected services between them, you can instantiate those views using this injected service:
```
using (var scope = _serviceScopeFactory.CreateScope())
{
    //Instantiate views
}
```

All views instantiated this way will share the same instance of injected scoped services. All you have to do is to inject IServiceScopeFactory.

**Reply by Xyncgas** [reply]
Scoped service is a easy way to create resources that would be garbage collected if navigated away saving memory
